lib/iris/experimental/ugrid.py

```diff
@@ -1680,102 +1679,316 @@ def remove(
+class _MeshConnectivityManagerMixin(ABC):
+    REQUIRED = ()
+    OPTIONAL = ()
+    NDIM = NotImplemented
+
+    @abstractmethod
+    def __init__(self, *connectivities):
+        cf_roles = [c.cf_role for c in connectivities]
+        for requisite in self.REQUIRED:
+            if requisite not in cf_roles:
+                message = f"{self.NDIM}D meshes require a {requisite}."
+                raise ValueError(message)
+
+        self.ALL = self.REQUIRED + self.OPTIONAL
+        self._members = {member: None for member in self.ALL}
+        self.add(*connectivities)
+
+    def __eq__(self, other):
+        # TBD
+        return NotImplemented
+
+    def __getstate__(self):
+        # TBD
+        pass
+
+    def __iter__(self):
+        for item in self._members.items():
+            yield item
+
+    def __ne__(self, other):
+        # TBD
+        return NotImplemented
+
+    def __repr__(self):
+        args = [
+            f"{member}={connectivity!r}"
+            for member, connectivity in self
+            if connectivity is not None
+        ]
+        return f"{self.__class__.__name__}({', '.join(args)})"
+
+    def __setstate__(self, state):
+        # TBD
+        pass
+
+    def __str__(self):
+        args = [
+            f"{member}=True"
+            for member, connectivity in self
+            if connectivity is not None
+        ]
+        return f"{self.__class__.__name__}({', '.join(args)})"
+
+    @property
+    @abstractmethod
+    def all_members(self):
+        return NotImplemented
+
+    def add(self, *connectivities):
+        # Since Connectivity classes include their cf_role, no setters will be
+        # provided, just a means to add one or more connectivities to the
+        # manager.
+        # No warning is raised for duplicate cf_roles - user is trusted to
+        # validate their outputs.
+        add_dict = {}
+        for connectivity in connectivities:
+            if not isinstance(connectivity, Connectivity):
+                message = f"Expected Connectivity, got: {type(connectivity)} ."
+                raise ValueError(message)
+            cf_role = connectivity.cf_role
+            if cf_role not in self.ALL:
+                message = (
+                    f"Not adding connectivity ({cf_role}: "
+                    f"{connectivity!r}) - cf_role must be one of: {self.ALL} ."
+                )
+                logger.debug(message, extra=dict(cls=self.__class__.__name__))
+            else:
+                add_dict[cf_role] = connectivity
+
+        # Validate shapes.
+        proposed_members = {**self._members, **add_dict}
+        locations = set(
+            [
+                c.src_location
+                for c in proposed_members.values()
+                if c is not None
+            ]
+        )
+        for location in locations:
+            counts = [
+                len(c.indices_by_src())
```

Comment:
@trexfeathers Just checking, but would this be an expensive check for large meshes?

Reply:
Thanks! I'd forgotten my own syntax for preserving laziness.

3ca9fa1e3cfaf5bb3709dcef1b23682c73235b2d